### PR TITLE
Fix: #6087 Display specific error message for course slug collision (revision)

### LINKED ISSUE #6087: Error message from course slug collision does not explain the problem
### What is happening?

Attempting to change a course slug component (such as title) will result in a server error if it changes the slug to match another course. However, the error message shown in the UI is just "Internal Server Error". It should instead explain the problem so that the user can choose a different string, or navigate to the other course and delete it (if it's not being used).

### PR DESCRIPTION
**Pull request by ichandrasharma**
## What this PR does
(Revision)
Fixes https://github.com/WikiEducationFoundation/WikiEduDashboard/issues/6087

This PR addresses the issue where a generic "Internal Server Error" message was displayed when a user attempted to change a course slug to one that already existed.

## Screenshots
Before:

This is an error occurring not showing proper message only showing Internal Server Error
Screenshot (69)

Course titles can be identical for different organizations or institutions, but must be unique within the same organization or institution. This is because course titles are used to dynamically generate URLs (slugs), and duplicate titles within the same organization would create URL conflicts.
look: below
Screenshot (70)

After:

Showing proper message
Updated
Screenshot (75)
Screenshot (76)

### CODE REVIEW

File: app/assets/javascripts/utils/api.js [hunk 1]
```diff
@@ -321,6 +321,10 @@ const API = {
     });
 
     if (!response.ok) {
+      if (response.status === 422 || response.status === 400) {
```
[the remark this answers is not in the document]
Comment (ichandrasharma) [reply app/assets/javascripts/utils/api.js]: @ragesoss I've made the changes you can check them. Now, the backend dynamically returns JSON error responses in all supported languages based on the selected language. I’ve also updated the screenshots for your review. Let me know if you have any feedback!

**Comment by JiyaGupta-cs** [comment]
Just a suggestion, Displaying the existing course slug as a clickable link opening that existing course in a new tab so that users can easily identify which course it is colliding with by clicking on it. This will enhance clarity and user experience.

**Comment by ichandrasharma** [comment]
>  Just a suggestion, Displaying the existing course slug as a clickable link opening that existing course in a new tab so that users can easily identify which course it is colliding with by clicking on it. This will enhance clarity and user experience.
> What do you think?

It can be implemented, but the primary goal is to provide a clear and accurate message so that the user can choose a different slug.

File: app/assets/javascripts/utils/api.js [hunk 2]
```diff
@@ -321,6 +321,11 @@ const API = {
     });
 
     if (!response.ok) {
+      if (response.status === 422 ) {
```
Comment (ragesoss) [inline app/assets/javascripts/utils/api.js:324]: Is this part of the change necessary? For other errors where we display the backend message, the existing data flows work without special handling of a specific status.

Comment (ichandrasharma) [reply app/assets/javascripts/utils/api.js:324]: A 422 response returns a JSON structure (data.errors) with a dynamically generated slug, while other responses return plain text. If special handle is not applied, it results in an empty object, failing to extract the error message and causing the default Sentry log error (e.g., "Sentry is not defined")

Comment (ragesoss) [reply app/assets/javascripts/utils/api.js:324]: Other cases that use this error message handling pattern don't require extra code for handling the specific status. Can you identify what is different about this, compared with (for example) `#createRandomPeerAssignments`? I'd like for the error handling flow to be as simple and uniform as possible, so that most errors rely on the same code pathways without introducing new frontend code for each new case that needs user-readable error messages.

Comment (ichandrasharma) [reply app/assets/javascripts/utils/api.js:324]: if i remove special handle this happens it throws empty [] this becomes default message of sentry
Screenshot (78)

File: app/controllers/courses_controller.rb [hunk 3]
```diff
@@ -44,13 +44,20 @@ def update
     validate
     handle_course_announcement(@course.instructors.first)
     slug_from_params if should_set_slug?
-    @course.update update_params
+    if @course.update update_params
```
[the remark this answers is not in the document]
Comment (ichandrasharma) [reply app/controllers/courses_controller.rb:47]: you are right i will remove

File: app/assets/javascripts/utils/api.js (hunk 2 above)
Comment (ichandrasharma) [reply app/assets/javascripts/utils/api.js]: @ragesoss The code is working perfectly after removing the special Handle logic and adding a try-catch block around Sentry.captureMessage to handle failures. Unnecessary if conditions have also been removed. Please let me know if you need any more refinements.

File: app/controllers/courses_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ichandrasharma) [reply app/controllers/courses_controller.rb]: Removed Unnecessary if block

File: app/controllers/courses_controller.rb [hunk 4]
```diff
@@ -51,6 +51,11 @@ def update
     ensure_passcode_set
     UpdateCourseWorker.schedule_edits(course: @course, editing_user: current_user)
     render json: { course: @course }
+  rescue Course::DuplicateCourseSlugError => e
+  message = I18n.t('courses.error.duplicate_slug', slug: e.slug)
```
Comment (ragesoss) [inline app/controllers/courses_controller.rb:55]: This is not indented properly. I expect linting to fail.

Comment (ichandrasharma) [reply app/controllers/courses_controller.rb]: @ragesoss I have fixed the linting issues, ensuring proper formatting and readability. Let me know if you need any further improvements

**Comment by ragesoss** [comment]
It looks like a test broke because of this new validation. You likely need to make a slight adjustment to that test so that it doesn't trigger the duplicate slug error.

**Comment by ichandrasharma** [comment]
> It looks like a test broke because of this new validation. You likely need to make a slight adjustment to that test so that it doesn't trigger the duplicate slug error.



This is before adding any tests
Screenshot (81)
Screenshot (82)


This is after adding tests
Screenshot (86)


This means below error or ruby rspec test suit fail is not because of my code. but for additional, I added tests as you said
Screenshot (87)


In the below, all three already merged pulls have the same error in building
Screenshot (88)
Screenshot (89)


This is happening because of different code in the codebase  or  May be I am wrong but I think this is happening because of version conflict why I am saying, look below pictures
Screenshot (90)
Screenshot (91)
Screenshot (92)

look in our codebase below the picture
Screenshot (93)

but we are using this
Screenshot (94)




<br>

Conclusion:
As you mentioned, I added tests and verified them. The Ruby RSpec test suite failure is likely due to differences in the codebase unrelated to my changes or possibly for the reasons I explained earlier with the provided screenshots. Let me know if you need further improvement.

**Comment by ragesoss** [comment]
Have a look at the failed test run for this PR: https://github.com/WikiEducationFoundation/WikiEduDashboard/actions/runs/13287364003/job/37124931915

It's different from the ones in your screenshots, and the error is the course duplicate slug error. Run that test locally and see if you can replicate the failure.

**Comment by ichandrasharma** [comment]
> Have a look at the failed test run for this PR: https://github.com/WikiEducationFoundation/WikiEduDashboard/actions/runs/13287364003/job/37124931915
> 
> It's different from the ones in your screenshots, and the error is the course duplicate slug error. Run that test locally and see if you can replicate the failure.



@ragesoss 
Yes, sir, you were right. The error was caused by the new validation for duplicate slug errors. I have fixed it and attached the screenshots. Additionally, I have added tests for the new validation. Please review and let me know if any further improvements are needed. Thanks!

before
Screenshot (96)
Screenshot (97)


after
Screenshot (98)

**Comment by ichandrasharma** [comment]
@ragesoss I have made changes please review and let me know if need any further improvement

File: spec/lib/revision_stat_spec.rb [hunk 5]
```diff
@@ -49,7 +49,10 @@
 
     context 'course id' do
       context 'not for this course' do
-        before { course.update(id: 1000) }
+        before do
+          allow_any_instance_of(Course).to receive(:check_duplicate_slug).and_return(true)
```
Comment (ragesoss) [inline spec/lib/revision_stat_spec.rb:53]: This test shouldn't require working around the validation; a better strategy would be to update it so that it doesn't trigger the error when the validation is run. Maybe that means including an updated slug along with the ID.

However, this test is pretty confusing and I don't think it does anything useful. Let's just remove it.

Comment (ichandrasharma) [reply spec/lib/revision_stat_spec.rb]: @ragesoss removed confusing tests please check and let me know any further improvement

Screenshot (101)

**Comment by ragesoss** [comment]
Thanks! On further reflection, I'm not sure that using a validation is the right strategy here, since we already enforce unique slugs at the database level. With the validation, we'll be making an extra query every time we call `Course#update`. I think a better strategy would be to just rescue the database error that occurs without the validation, rather than rescuing an error thrown by the validation.
